hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieReaderContext.java

```diff
@@ -262,30 +265,52 @@ private BiFunction<T, Schema, String> virtualKeyExtractor(String[] recordKeyFiel
   /**
    * Gets the ordering value in particular type.
    *
-   * @param record An option of record.
+   * @param record An engine specific record.
    * @param schema The Avro schema of the record.
-   * @param orderingFieldName name of the ordering field
+   * @param orderingFieldName name of the ordering field, if any
    * @return The ordering value.
    */
-  public Comparable getOrderingValue(T record,
-                                     Schema schema,
-                                     Option<String> orderingFieldName) {
+  public Comparable getOrderingValue(T record, Schema schema, Option<String> orderingFieldName) {
     if (orderingFieldName.isEmpty()) {
       return DEFAULT_ORDERING_VALUE;
     }
 
     Object value = getValue(record, schema, orderingFieldName.get());
-    Comparable finalOrderingVal = value != null ? convertValueToEngineType((Comparable) value) : DEFAULT_ORDERING_VALUE;
-    return finalOrderingVal;
+    return value != null ? convertValueToEngineType((Comparable) value) : DEFAULT_ORDERING_VALUE;
   }
 
   /**
    * Constructs a new {@link HoodieRecord} based on the given buffered record {@link BufferedRecord}.
+   * Safe to assume that the buffered record is not a delete.
    *
    * @param bufferedRecord  The {@link BufferedRecord} object with engine-specific row
    * @return A new instance of {@link HoodieRecord}.
    */
-  public abstract HoodieRecord<T> constructHoodieRecord(BufferedRecord<T> bufferedRecord);
+  protected abstract HoodieRecord<T> constructHoodieDataRecord(BufferedRecord<T> bufferedRecord);
```

Comment:
I don't think this makes sense, we still need to keep the delete payloads for streaming retraction scenarios.

Reply:
Can you link to these scenarios so I can get a better understanding of how they are used?

Reply:
For example, the schema is (`id`, `name`, `val`) and the `id` is the record key, there is an insert then update to the value of it, you have msg events like below and the operator sends one msg at a time to the downstream:

```java
[+I] [1, "a", 1]
[-U] [1, "a", 1]
[+U] [1, "a", 3]
```

The `-U` msg is a retraction msg to downstream, when the downstream operators received the msg, it would minus the current value `1` with `1` so it becomes `0`.

The point here is to keep the data payload of the delete msg so the downstream can figure out the value to subtract.

Reply:
Ok, seems like this is related to the question [here](https://github.com/apache/hudi/pull/13242/files/3a8a4f0db770ece4a51aabff7d999e174788f017#r2091645480)